### PR TITLE
Support directory listing

### LINKED ISSUE #222: Support listing files in a directory Path
Listing files within a directory is one of the few file APIs my library tests require that `kotlinx-io` doesn't currently support.

Similar API references:

* [JVM File.listFiles](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#listFiles--)
* [Apple Foundation NSFileManager.contentsOfDirectoryAtPath:error:](https://developer.apple.com/documentation/foundation/nsfilemanager/1414584-contentsofdirectoryatpath)
* [Okio FileSystem.list](https://square.github.io/okio/3.x/okio/okio/okio/-file-system/list.html)

### PR DESCRIPTION
**Pull request by fzhinkin**
It'll take a while to implement proper FS API (#241), so it does make sense to provide some API for listing directories for now.

Closes #222

### CODE REVIEW

File: core/common/src/files/FileSystem.kt [hunk 1]
```diff
@@ -145,6 +145,20 @@ public sealed interface FileSystem {
...
+     * If path [directory] was an absolute path, a returned list will also contain absolute paths.
+     * If it was a relative path, a returned list will contain relative paths.
+     *
+     * @param directory a directory to list.
+     * @return a list of [directory]'s immediate children.
+     * @throws FileNotFoundException if [directory] does not exist.
+     * @throws IOException if [directory] points to something other than directory.
+     * @throws IOException if there was an underlying error preventing listing [directory] children.
+     */
+    public fun list(directory: Path): List<Path>
```
Comment (qwwdfsad) [inline core/common/src/files/FileSystem.kt:161]: When looking at other ways to list the directory, folks typically avoid aggregating it in a single list, I suspect for multiple reasons:

* `opendir_at` and ability to be consistent when operating with a collection of paths (also, resolve 
* There are concerns w.r.t. performance and footprint: it is, in fact, a bulk operation that cannot be reasonably limited by the caller (and doesn't make them think to do so)
* [Kotlin-specific] "Lazy, iterable, closeable and countable sequence" is a core primitive for the filesystem, and `list` can be expressed in terms of sequence, while the opposite is not true

Comment (qwwdfsad) [reply core/common/src/files/FileSystem.kt:161]: Probably worth considering a core methods returning a sequence and `toList` as a way to materialize it

Comment (fzhinkin) [reply core/common/src/files/FileSystem.kt:161]: Then we need a closable sequence (which is not a thing, unlike Java's Stream).

Comment (qwwdfsad) [reply core/common/src/files/FileSystem.kt:161]: There are other alternatives, such as `FS.forEachEntry(dir) { e -> }`.

My point was not to propose another solution but to highlight the potential pitfalls and point out that in other places, folks follow a different path [probably for a reason].
It's totally okay if you considered all of them and disregarded

File: core/common/test/files/SmokeFileTest.kt [hunk 2]
```diff
@@ -443,6 +443,36 @@ class SmokeFileTest {
...
+        assertFailsWith<IOException> { SystemFileSystem.list(tmpFile) }
+
+        val dir = createTempPath().also {
+            SystemFileSystem.createDirectories(it)
+        }
+        assertEquals(emptyList(), SystemFileSystem.list(dir))
+
+        val subdir = Path(dir, "subdir").also {
+            SystemFileSystem.createDirectories(it)
+            SystemFileSystem.sink(Path(it, "file")).close()
```
Comment (shanshin) [inline core/common/test/files/SmokeFileTest.kt:462]: Why we create file here?

File: core/common/test/files/SmokeFileTest.kt [hunk 3]
```diff
@@ -443,6 +443,36 @@ class SmokeFileTest {
...
+        }
+        assertEquals(listOf(subdir), SystemFileSystem.list(dir))
+
+        val file = Path(dir, "file").also {
+            SystemFileSystem.sink(it).close()
```
Comment (shanshin) [inline core/common/test/files/SmokeFileTest.kt:467]: Why not test other file types, such as symlinks?

File: core/common/src/files/FileSystem.kt (hunk 1 above)
Comment (shanshin) [reply core/common/src/files/FileSystem.kt:161]: It seems to me that ideally there should be something like what is in stdlib, but I'm not sure that this needs to be done right now.
Is it worth adding an API now that we know in advance that we will remove it soon?!

File: build-logic/src/main/kotlin/kotlinx/io/conventions/kotlinx-io-multiplatform.gradle.kts [hunk 4]
```diff
@@ -80,6 +80,12 @@ kotlin {
                         group("androidNative")
                     }
                 }
+
+                group("nonAndroid") {
```
Comment (shanshin) [inline build-logic/src/main/kotlin/kotlinx/io/conventions/kotlinx-io-multiplatform.gradle.kts:84]: Am I right that `nonAndroid` is now actually a synonym for `posix`?

File: core/common/test/files/SmokeFileTest.kt (hunk 3 above)
Comment (fzhinkin) [reply core/common/test/files/SmokeFileTest.kt:467]: We don't support symlinks yet (in wasmWasi it was supported only to make other things work correctly).

File: core/common/test/files/SmokeFileTest.kt (hunk 2 above)
Comment (fzhinkin) [reply core/common/test/files/SmokeFileTest.kt:462]: To ensure it won't be listed (i.e. that `list` is not recursive).

File: core/common/test/files/SmokeFileTest.kt (hunk 3 above)
Comment (shanshin) [reply core/common/test/files/SmokeFileTest.kt:467]: Symlinks can't be listed from directory?

File: core/common/src/files/FileSystem.kt (hunk 1 above)
Comment (fzhinkin) [reply core/common/src/files/FileSystem.kt:161]: >  soon?!

It depends on how you perceive time, I'm afraid.

File: build-logic/src/main/kotlin/kotlinx/io/conventions/kotlinx-io-multiplatform.gradle.kts (hunk 4 above)
Comment (fzhinkin) [reply build-logic/src/main/kotlin/kotlinx/io/conventions/kotlinx-io-multiplatform.gradle.kts:84]: Well, the only POSIX certified target down there is macOs :)

File: core/common/test/files/SmokeFileTest.kt (hunk 3 above)
Comment (fzhinkin) [reply core/common/test/files/SmokeFileTest.kt:467]: We don't have an API to create them for testing purposes. They can be listed, of course.

File: build-logic/src/main/kotlin/kotlinx/io/conventions/kotlinx-io-multiplatform.gradle.kts (hunk 4 above)
Comment (shanshin) [reply build-logic/src/main/kotlin/kotlinx/io/conventions/kotlinx-io-multiplatform.gradle.kts:84]: But for somehow we use posix API for all of them ))

Comment (shanshin) [reply build-logic/src/main/kotlin/kotlinx/io/conventions/kotlinx-io-multiplatform.gradle.kts:84]: It's just that for me, JVM is also a non-Android, so this name was confused at first

Comment (fzhinkin) [reply build-logic/src/main/kotlin/kotlinx/io/conventions/kotlinx-io-multiplatform.gradle.kts:84]: But Android is as POSIX as Linux

Comment (fzhinkin) [reply build-logic/src/main/kotlin/kotlinx/io/conventions/kotlinx-io-multiplatform.gradle.kts:84]: Perhaps, `commonizedAsPosix`?

File: core/wasmWasi/test/WasiFsTest.kt [hunk 5]
```diff
@@ -219,4 +219,14 @@ class WasiFsTest {
...
         }
     }
+
+    // https://github.com/nodejs/node/blob/6f4d6011ea1b448cf21f5d363c44e4a4c56ca34c/deps/uvwasi/src/uvwasi.c#L19
+    @Test
+    fun readdirUnsupportedOnWindows() {
+        if (!isWindows) return
+
+        assertFailsWith<IOException> {
+            SystemFileSystem.list(PreOpens.roots.first())
```
[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/wasmWasi/test/WasiFsTest.kt:229]: I don't see how we can overcome this issue. Perhaps, the best thing we can do is explicitly document that directory listing does not work for wasmWasi target in conjunction with NodeJs on Windows.

File: build-logic/src/main/kotlin/kotlinx/io/conventions/kotlinx-io-multiplatform.gradle.kts (hunk 4 above)
Comment (fzhinkin) [reply build-logic/src/main/kotlin/kotlinx/io/conventions/kotlinx-io-multiplatform.gradle.kts:84]: Yet another strenuous attempt to give a better name:
- `nativeNonAndroid`
- rename `nonApple` to `nativeNonApple`

WDYT?

File: core/wasmWasi/test/WasiFsTest.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/wasmWasi/test/WasiFsTest.kt:229]: Opened https://github.com/Kotlin/kotlinx-io/issues/312

Also, I'll document lack of support on NodeJs/Windows

File: core/common/test/files/SmokeFileTest.kt (hunk 3 above)
Comment (shanshin) [reply core/common/test/files/SmokeFileTest.kt:467]: Issue https://github.com/Kotlin/kotlinx-io/issues/241

File: core/wasmWasi/test/WasiFsTest.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/wasmWasi/test/WasiFsTest.kt:229]: Done

**Comment by StefanOltmann** [comment]
Nice! :)

**Pull request merged**